massa-versioning-worker/src/versioning_factory.rs

```diff
@@ -0,0 +1,402 @@
+    #[error("Version: {0} is not active yet")]
+    OnStateNotReady(u32),
+    #[error("Could not create object of type {0}: {1}")]
+    OnCreate(String, String),
+}
+
+#[derive(Clone)]
+/// Strategy to use when creating a new object from a factory
+pub enum FactoryStrategy {
+    /// use get_best_version (see Factory trait)
+    Best,
+    /// Require to create an object with this specific version
+    Exact(u32),
+    /// Create an object given a timestamp (e.g slot)
+    At(MassaTime),
+}
+
+impl From<u32> for FactoryStrategy {
+    fn from(value: u32) -> Self {
+        FactoryStrategy::Exact(value)
+    }
+}
+
+// Factory traits
+
+/// Trait for Factory that create objects based on versioning store
+pub trait VersioningFactory {
+    /// Factory will create only one object of a given type,
+    /// so a FactoryAddress (impl VersioningFactory), will have type Output = Address
+    type Output;
+    /// Error if object cannot be created (+ used for VersioningFactoryArgs)
+    type Error;
+    /// Arguments struct in order to create Self::Output
+    type Arguments;
+
+    /// Return the MipComponent associated with return type (Output)
+    /// e.g. if type Output = Address, should return MipComponent::Address
+    fn get_component() -> MipComponent;
+    /// Access to the MipStore
+    fn get_versioning_store(&self) -> MipStore;
+
+    /// Get best version (aka last active for factory component)
+    fn get_best_version(&self) -> u32 {
+        let component = Self::get_component();
+        let vi_store_ = self.get_versioning_store();
+        let vi_store = vi_store_.0.read();
+        let state_active = MipState::active();
+
+        vi_store
+            .0
+            .iter()
+            .rev()
+            .find_map(|(vi, vsh)| {
+                (vi.component == component && vsh.state == state_active)
+                    .then_some(vi.component_version)
+            })
+            .unwrap_or(0)
+    }
+
+    /// Get best version at given timestamp (e.g. slot)
+    fn get_best_version_at(&self, ts: MassaTime) -> Result<u32, FactoryError> {
+        let component = Self::get_component();
+        let vi_store_ = self.get_versioning_store();
+        let vi_store = vi_store_.0.read();
+        let state_active = MipState::active();
+
+        // Iter backward, filter component + state active,
+        let version = vi_store
+            .0
+            .iter()
+            .rev()
+            .filter(|(vi, vsh)| vi.component == component && vsh.state == state_active)
+            .find_map(|(vi, vsh)| {
+                let res = vsh.state_at(ts, vi.start, vi.timeout);
+                match res {
+                    Ok(MipStateTypeId::Active) => Some(vi.component_version),
+                    _ => None,
+                }
+            })
+            .unwrap_or(0);
+
+        Ok(version)
+    }
+
+    /// Get all versions in 'Active state' for the associated MipComponent
+    fn get_versions(&self) -> Vec<u32> {
```

Comment:
As you have `get_all_versions` below, I think it's clearer if you have `get_all_active_versions` here

Reply:
Again this add some 'State info' to the dev, let's talk about it on Monday :)

Reply:
Your concern is valid, let's try to improve the function naming without "manually" exposing the state we are in. We should also desambiguate the "network version" and the "component version" as much as possible :)

Reply:
done